core/src/main/java/org/elasticsearch/common/cache/Cache.java

```diff
@@ -0,0 +1,667 @@
+                        throw new ExecutionException(new NullPointerException("loader returned a null value"));
+                    }
+                    put(key, value, now);
+                }
+            }
+        }
+        return value;
+    }
+
+    /**
+     * Associates the specified value with the specified key in this map. If the map previously contained a mapping for
+     * the key, the old value is replaced.
+     *
+     * @param key   key with which the specified value is to be associated
+     * @param value value to be associated with the specified key
+     */
+    public void put(K key, V value) {
+        long now = now();
+        put(key, value, now);
+    }
+
+    private void put(K key, V value, long now) {
+        CacheSegment<K, V> segment = getCacheSegment(key);
+        Tuple<Entry<K, V>, Entry<K, V>> tuple = segment.put(key, value, now);
+        boolean replaced = false;
+        try (ReleasableLock ignored = lruLock.acquire()) {
+            if (tuple.v2() != null && tuple.v2().state == State.EXISTING) {
+                if (unlink(tuple.v2())) {
+                    replaced = true;
+                }
+            }
+            promote(tuple.v1(), now);
+        }
+        if (replaced) {
+            removalListener.onRemoval(new RemovalNotification(tuple.v2().key, tuple.v2().value, RemovalNotification.RemovalReason.REPLACED));
+        }
+    }
+
+    /**
+     * Invalidate the association for the specified key. A removal notification will be issued for invalidated
+     * entries with {@link org.elasticsearch.common.cache.RemovalNotification.RemovalReason} INVALIDATED.
+     *
+     * @param key the key whose mapping is to be invalidated from the cache
+     */
+    public void invalidate(K key) {
+        CacheSegment<K, V> segment = getCacheSegment(key);
+        Entry<K, V> entry = segment.remove(key);
+        if (entry != null) {
+            try (ReleasableLock ignored = lruLock.acquire()) {
+                delete(entry, RemovalNotification.RemovalReason.INVALIDATED);
+            }
+        }
+    }
+
+    /**
+     * Invalidate all cache entries. A removal notification will be issued for invalidated entries with
+     * {@link org.elasticsearch.common.cache.RemovalNotification.RemovalReason} INVALIDATED.
+     */
+    public void invalidateAll() {
+        Entry<K, V> h = head;
+        Arrays.stream(segments).forEach(segment -> segment.segmentLock.writeLock().lock());
+        try (ReleasableLock ignored = lruLock.acquire()) {
+            Arrays.stream(segments).forEach(segment -> segment.map = new HashMap<>());
+            Entry<K, V> current = head;
+            while (current != null) {
+                current.state = State.DELETED;
+                current = current.after;
+            }
+            head = tail = null;
+            count = 0;
+            weight = 0;
+        }
+        Arrays.stream(segments).forEach(segment -> segment.segmentLock.writeLock().unlock());
```

Comment:
I've discussed this (offline) with a few people. I'm not sure if there are any scenarios where we can take the locks, an exception can be thrown on the side, and it matters if the locks are not released (e.g., we could OOME, but then the process is coming down anyway).

But, I understand doing it for sanity. Can you check 881593eea9de0c6d6d726c163996bb387081adb0 carefully?